### PR TITLE
[SPARK-24341][SQL][followup] remove duplicated error checking

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

There are 2 places we check for problematic `InSubquery`: the rule `ResolveSubquery` and `InSubquery.checkInputDataTypes`. We should unify them.

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by dilipbiswal** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by dilipbiswal** [comment]
@cloud-fan oops... sorry.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 1]
```diff
@@ -155,37 +155,35 @@ case class InSubquery(values: Seq[Expression], query: ListQuery)
...
-          s"""
-             |The data type of one or more elements in the left hand side of an IN subquery
-             |is not compatible with the data type of the output of the subquery
-             |Mismatched columns:
-             |[${mismatchedColumns.mkString(", ")}]
-             |Left side:
-             |[${values.map(_.dataType.catalogString).mkString(", ")}].
-             |Right side:
-             |[${query.childOutputs.map(_.dataType.catalogString).mkString(", ")}].""".stripMargin)
+    if (values.length != query.childOutputs.length) {
```
[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: the reason why I added the check in `ResolveSubquery` was to fail fast immediately, without wasting time doing all the analysis until we check the data types. In this way, the check in  `checkInputDataTypes` was actually useless, but I left it there for safety.

I agree unifying them and actually I am fine with this change, but I'd prefer keeping the check in `ResolveSubquery` in order to fail immediately (it is not a big deal anyway). What do you think?

[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: That's not how it's designed to work. The reason we have `CheckAnalysis` is to do analysis work as much as we can, and report the most precise error at the end. Performance is not a big concern in analyzer, giving users the best error message is more important. Although in this case the error message is not a problem, it's better to follow the designed way.

[the remark this answers is not in the document]
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: Yes, thanks.

**Comment by mgaido91** [comment]
retest this please

**Comment by mgaido91** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**